From 627af8c8165d7fa4f28a3c23b2ce7a9c1096fc0b Mon Sep 17 00:00:00 2001
From: Marton Balint <cus@passwd.hu>
Date: Sat, 13 Mar 2021 22:33:38 +0100
Subject: [PATCH] avcodec/assdec: do not overread if zero padding is missing

Fixes: a3ewdvDE2QSe
Fixes: hVJyTegWPmu6
Fixes: out of array read
Fixes: wrong event text
Out of array read Replicated through API with ASAN
Wrong event text Replicated through UnModified FFmpeg with plain output diff

Signed-off-by: Marton Balint <cus@passwd.hu>
(cherry picked from commit 627af8c8165d7fa4f28a3c23b2ce7a9c1096fc0b)
---
 libavcodec/assdec.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavcodec/assdec.c b/libavcodec/assdec.c
index 89d7b51894b..28aebd4d0e6 100644
--- a/libavcodec/assdec.c
+++ b/libavcodec/assdec.c
@@ -54,7 +54,7 @@ static int ass_decode_frame(AVCodecContext *avctx, AVSubtitle *sub,
         return AVERROR(ENOMEM);
     sub->num_rects = 1;
     sub->rects[0]->type = SUBTITLE_ASS;
-    sub->rects[0]->ass  = av_strdup(avpkt->data);
+    sub->rects[0]->ass  = av_strndup(avpkt->data, avpkt->size);
     if (!sub->rects[0]->ass)
         return AVERROR(ENOMEM);
     *got_sub_ptr = 1;
-- 
2.53.0

